mirgecom/navierstokes.py

```diff
@@ -0,0 +1,416 @@
+        discr, dd_vol_quad, dd_faces_quad, vol_state_quad.temperature, t_flux_bnd)
+
+
+def ns_operator(discr, gas_model, state, boundaries, *, time=0.0,
+                inviscid_numerical_flux_func=inviscid_flux_rusanov,
+                gradient_numerical_flux_func=gradient_flux_central,
+                viscous_numerical_flux_func=viscous_flux_central,
+                quadrature_tag=DISCR_TAG_BASE,
+                # Added to avoid repeated computation
+                # FIXME: See if there's a better way to do this
+                operator_states_quad=None,
+                grad_cv=None, grad_t=None):
+    r"""Compute RHS of the Navier-Stokes equations.
+
+    Parameters
+    ----------
+    state: :class:`~mirgecom.gas_model.FluidState`
+
+        Fluid state object with the conserved state, and dependent
+        quantities.
+
+    boundaries
+        Dictionary of boundary functions keyed by btags
+
+    time
+        Time
+
+    gas_model: :class:`~mirgecom.gas_model.GasModel`
+
+        Physical gas model including equation of state, transport,
+        and kinetic properties as required by fluid state
+
+    quadrature_tag
+        An identifier denoting a particular quadrature discretization to use during
+        operator evaluations.
+
+    Returns
+    -------
+    :class:`mirgecom.fluid.ConservedVars`
+
+        The RHS of the Navier-Stokes equations.
+
+        .. math::
+
+            \partial_t \mathbf{Q} = \nabla\cdot(\mathbf{F}_V - \mathbf{F}_I)
+    """
+    if not state.is_viscous:
+        raise ValueError("Navier-Stokes operator expects viscous gas model.")
+
+    dd_base = as_dofdesc("vol")
+    dd_vol_quad = DOFDesc("vol", quadrature_tag)
+    dd_faces_quad = DOFDesc("all_faces", quadrature_tag)
+
+    # Make model-consistent fluid state data (i.e. CV *and* DV) for:
+    # - Volume: vol_state_quad
+    # - Element-element boundary face trace pairs: inter_elem_bnd_states_quad
+    # - Interior states (Q_minus) on the domain boundary: domain_bnd_states_quad
+    #
+    # Note: these states will live on the quadrature domain if one is given,
+    # otherwise they stay on the interpolatory/base domain.
+    if operator_states_quad is None:
+        operator_states_quad = make_operator_fluid_states(
+            discr, state, gas_model, boundaries, quadrature_tag)
+
+    vol_state_quad, inter_elem_bnd_states_quad, domain_bnd_states_quad = \
+        operator_states_quad
+
+    # {{{ Local utilities
+
+    # transfer trace pairs to quad grid, update pair dd
+    def _interp_to_surf_quad(utpair):
+        local_dd = utpair.dd
+        local_dd_quad = local_dd.with_discr_tag(quadrature_tag)
+        return TracePair(
+            local_dd_quad,
+            interior=op.project(discr, local_dd, local_dd_quad, utpair.int),
+            exterior=op.project(discr, local_dd, local_dd_quad, utpair.ext)
+        )
+
+    # }}}
+
+    # {{{ === Compute grad(CV) ===
+
+    if grad_cv is None:
+        grad_cv = grad_cv_operator(
```

Comment:
Why are there separate gradient functions for CV and temparature? A gradient is a gradient, right?

Reply:
The two functions interact with the fluid state and boundary conditions differently. It could see it being possible to create a single `gradient_operator` that could be used for both, but that's probably a story for another PR...

Reply:
These are the CV-specific and fluid-temperature-specific handlers to compute the gradients of those quantities.  Underneath, they both use the gradient-is-a-gradient interface `gradient_operator`.